scan/utils.go

```diff
@@ -18,14 +18,142 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 package scan
 
 import (
+	"encoding/json"
 	"fmt"
 	"strings"
 
 	"github.com/future-architect/vuls/config"
 	"github.com/future-architect/vuls/models"
 	"github.com/future-architect/vuls/util"
+	"github.com/k0kubun/pp"
+	"time"
 )
 
+func Wpscan() (err error) {
+	if err = wpscanVuln(); err != nil {
+		return err
+	}
+	return
+}
+
+func wpscanVuln() (err error) {
+	errChan := make(chan error, len(config.Conf.Servers))
+	defer close(errChan)
+	for _, s := range config.Conf.Servers {
+		go func(s config.ServerInfo) {
+			defer func() {
+				if p := recover(); p != nil {
+					util.Log.Debugf("Panic: %s on %s", p, s.ServerName)
+				}
+			}()
+			errChan <- detectWp(s)
+		}(s)
+	}
+
+	timeout := time.After(time.Duration(10000) * time.Second)
+	for i := 0; i < len(config.Conf.Servers); i++ {
+		select {
+		case _ = <-errChan:
+		/*
+			if 0 < len(res.getErrs()) {
+				errServers = append(errServers, res)
+				util.Log.Errorf("(%d/%d) Failed: %s, err: %s",
+					i+1, len(config.Conf.Servers),
+					res.getServerInfo().ServerName,
+					res.getErrs())
+			} else {
+				servers = append(servers, res)
+				util.Log.Infof("(%d/%d) Detected: %s: %s",
+					i+1, len(config.Conf.Servers),
+					res.getServerInfo().ServerName,
+					res.getDistro())
+			}
+			*/
+		case <-timeout:
+			msg := "Timed out while wordpress scan"
+			util.Log.Error(msg)
+			for servername, sInfo := range config.Conf.Servers {
+				found := false
+				for _, o := range append(servers, errServers...) {
+					if servername == o.getServerInfo().ServerName {
+						found = true
+						break
+					}
+				}
+				if !found {
+					u := &unknown{}
+					u.setServerInfo(sInfo)
+					u.setErrs([]error{
+						fmt.Errorf("Timed out"),
+					})
+					errServers = append(errServers, u)
+					util.Log.Errorf("(%d/%d) Timed out: %s",
+						i+1, len(config.Conf.Servers),
+						servername)
+					i++
+				}
+			}
+		}
+	}
+	return
+}
+
+type Hoge struct {
+	Version Huga `json:"-"`
+}
+
+
+type Huga struct {
+	ReleaseDate 	string `json:"releaseDate"`
+	ChangelogUrl 	string `json:"changelogUrl"`
```

Comment:
__[Golint]__ struct field ChangelogUrl should be ChangelogURL

<sub>posted by Sider</sub>